packages/grafana-data/src/types/datasource.ts

```diff
@@ -285,12 +285,12 @@ abstract class DataSourceApi<
   /**
    * Get tag keys for adhoc filters
    */
-  getTagKeys?(options?: DataSourceGetTagKeysOptions<TQuery>): Promise<MetricFindValue[]>;
+  getTagKeys?(options?: DataSourceGetTagKeysOptions<TQuery>): Promise<GetTagResponse> | Promise<MetricFindValue[]>;
 
   /**
    * Get tag values for adhoc filters
    */
-  getTagValues?(options: DataSourceGetTagValuesOptions<TQuery>): Promise<MetricFindValue[]>;
+  getTagValues?(options: DataSourceGetTagValuesOptions<TQuery>): Promise<GetTagResponse> | Promise<MetricFindValue[]>;
```

Comment:
It should be ok to make this change. It is a breaking change for potential consumers of this function since they can start to receive an object instead of an array. However, I can't see many plugins calling this function but rather implementing it to provide support for template variables.

So I think that would be OK if we just add something to the breaking changes documentation.

An alternative approach be to add two new functions where we provide a default implementation which will check if the plugin implements the new function. If it does then we call it otherwise we call the existing function and wrap the return statement in an object so it looks like the `GetTagResponse`? But I'm not sure it is worth it. 

Something like:

```ts
async getVerboseTagKeys(options?: DataSourceGetTagKeysOptions<TQuery>): Promise<GetTagResponse> {
   if(isFunction(this.getVerboseTagKeys)) {
      return this.getVerboseTagKeys(options);
   }
   const keys = await this.getTagKeys(options);
   return { data: keys, errors: [] };
}
```

Reply:
This could work, but I think I prefer to keep the data source API slim and not add more methods.

Reply:
I agree with @sunker , would rather not introduce new methods. 

From the breaking changes POV this should not have an impact on plugins, this API is consumed by our internals only, I doubt it's used directly by plugins.

> Would be nice to include a code example in the PR description that demonstrates how the new response can be used in a data source.

Yes, that would be useful @kaydelaney